napari/_vispy/layers/labels.py

```diff
@@ -143,37 +219,87 @@ def hash2d_get(key, keys, values, empty_val=0):
     """
     pos = idx_to_2D(key, keys.shape)
     initial_key = key
-    while keys[pos] != key and keys[pos] != empty_val:
+    while keys[pos] != initial_key and keys[pos] != empty_val:
         if key - initial_key > keys.size:
             raise KeyError('label does not exist')
         key += 1
         pos = idx_to_2D(key, keys.shape)
-    return pos if keys[pos] == key else None
+    return pos if keys[pos] == initial_key else None
 
 
-def hash2d_set(key, value, keys, values, empty_val=0):
+def hash2d_set(key: float, value, keys, values, empty_val=0) -> bool:
     """
     Set a value in the 2d hashmap, wrapping around to avoid collision.
     """
-    if key is None:
-        return
+    if key is None or isnan(key):
+        return False
     pos = idx_to_2D(key, keys.shape)
     initial_key = key
+    collision = False
     while keys[pos] != empty_val:
+        collision = True
         if key - initial_key > keys.size:
             raise OverflowError('too many labels')
         key += 1
         pos = idx_to_2D(key, keys.shape)
-    keys[pos] = key
+    keys[pos] = initial_key
     values[pos] = value
 
+    return collision
 
-def build_textures_from_dict(color_dict, empty_val=0, shape=(1000, 1000)):
+
+def _get_shape_from_dict(color_dict):
+    size = len(color_dict) / MAX_LOAD_FACTOR
+    # I think that hash table size should be at least 4 times
+    # bigger than the number of labels to avoid collisions
```

Comment:
A comment explaining what load factor and prime num table are would be great here!

Reply:
could you check current comment?

Reply:
> If that was meant to be a docstring, it should be moved one line up.

No. I do not have idea to make it docstring, but maybe I should? 




> So I would expect a load factor > 1.3? you mean that it's basically 4 since you divide by 0.25? Either way, it's not super clear to me why you think a bigger value than literature here is a better idea.

The problem here is cost of collisions and cost of branched calculation on GPU. When you have factor 1.3 (0,76) there are some collisions, but on CPU they are much cheaper as there is no worker group that work in synchronization. 

> Do you mean that each subsequent element is roughly twice the previous? And that's good why?

First shader will be 61x61, second 61x127, third 127x127 etc. there is a need to made some decision. Doubling size looks good for me (except first one the shader will be maximum 8 times bigger than labels number).

Reply:
> No. I do not have idea to make it docstring, but maybe I should?

Maybe yes, I think it's better.

> The problem here is cost of collisions and cost of branched calculation on GPU. When you have factor 1.3 (0,76) there are some collisions, but on CPU they are much cheaper as there is no worker group that work in synchronization.

Ok, so basically compared to the "ideal" factor, we prefer over-using memory to gain a bit fewer collisions, because we prefer to maximize rendering performance instead of memory footprint?

> First shader will be 61x61, second 61x127, third 127x127 etc. there is a need to made some decision. Doubling size looks good for me (except first one the shader will be maximum 8 times bigger than labels number).

Ok... I guess this is a reasonable heurisitic, I don't have an intuition for what's best here :P

Reply:
I reimplement it to reduce collision. Could you check now?